### PR TITLE
License, Windows and Compatibility Upgrade (redux)

### PR DESCRIPTION
**Pull request by neurolabusc**
This pull request responds to [feedback](https://github.com/cloudflare/zlib/pull/18) to an earlier pull request.

It adds 3 benefits:
 1. The GPL code for SIMD CRC has been replaced by BSD-license code from the Chrome browser. Performance is identical, as demonstrated in this [minimal project](https://github.com/neurolabusc/simd_crc). This addresses [complications raised with GPL code](https://discourse.vtk.org/t/cloudflare-zlib/2330/13).
 2. At run-time, the software now detects if the host computer supports CLMUL instructions, using the accelerated CRC if supported. This prevents the software from executing illegal instructions on hardware that does not support CLMUL or where CLMUL has been disabled by a [OEM BIOS or virtual machine](https://github.com/InsightSoftwareConsortium/ITK/issues/416). The code for detection is designed to be backward compatible with old compilers (e.g. gcc 4.8.4) that report to CMAKE that they support CLMUL instructions but do not provide the cpuid detection of CLMUL. 
 3. The CMAKE files have been updated to [support modern Microsoft compilers](https://github.com/ningfei/zlib) allowing simpler distribution to those using the Windows operating system.

I would be excited to see this pull request included, as it can be called from the [pigz make](https://github.com/madler/pigz/issues/62). This dramatically accelerates pigz, with the SIMD CRC directly accelerating the serial stage that is most impacted by Amdahl's law.

The first time the CRC is called it set the atomic int cpu_has_pclmul, which reports whether or not the clmul instructions are available. This ensures the program does not crash if compiled on a modern computer and executed on an old computer. I appreciate different people have different views on this, which I summarize below. I do not have a preference, the pull request I am providing is robust (tested on Ubuntu 14.04 as well as recent versions) and simple. The CRC SIMD code and CPUID is only tested for x86-64, but again this extends the current repository. Future developments my consider alternatives:

 - @Adenilson suggests [pthread_once_t](https://github.com/madler/zlib/pull/468), though @mscastanho criticized this approach.
 - @mscastanho prefers [ifunc](https://github.com/madler/zlib/pull/468) but @Adenilson was concerned that the performance is not great and it is not available on all platforms.
 - @vkrasnov suggested [__attribute__((constructor))](https://github.com/cloudflare/zlib/pull/18) but @kornelski was not  sure if that was the right approach.
 

Other comments
 
 - @Adenilson suggested [SIMD CRC for ARM](https://github.com/cloudflare/zlib/pull/18) but I am unable to validate that platform.
 - The compiler [-mpclmul](https://github.com/jtkukunas/zlib/pull/25) flag is only applied to the crc32_simd unit. In theory, if this flag were applied globally, a compiler could optimize other sections of code to use these instructions, and the resulting code would fail if run on older (<2009) computers.

 For most cases (where files are large) testing the CPU for CLMUL support on has negligible performance impact. One can conceive of use cases where an statically compiled executable is called many times to compress tiny files. If one is confident the resulting executable is going to be only run on modern (Westmere or later) CPUs, one can compile with the DSKIP_CPUID_CHECK compiler directive to skip this check for cmake the command is "cmake -DSKIP_CPUID_CHECK=ON ..".
 
 - Single threaded performance [better than other zlib libraries](https://github.com/neurolabusc/zlib-bench)
 - Multi-threaded preformance [better than other zlib libraries](https://github.com/neurolabusc/pigz-bench)

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -762,15 +762,19 @@ EOF
...
   fi
-
-  # Check for PCLMUL support
+#Project copied from zlib-ng:
+# Check for PCLMUL support
 cat > $test.c << EOF
 #include <immintrin.h>
-void foo(void) {
-  _mm_clmulepi64_si128(_mm_set1_epi16(1), _mm_set1_epi16(2), 0);
+#include <wmmintrin.h>
```
Comment (vkrasnov) [inline configure]: this include is not required AFAIK

File: crc32.c [hunk 2]
```diff
@@ -21,6 +21,14 @@
   DYNAMIC_CRC_TABLE and MAKECRCH can be #defined to write out crc32.h.
  */
 
+#ifdef HAS_PCLMUL
+ #include "crc32_simd.h"
+ #ifndef _MSC_VER
+  #include <cpuid.h>
+ #endif
+ //#include <stdio.h>
```
Comment (vkrasnov) [inline crc32.c]: please remove

File: crc32.c [hunk 3]
```diff
@@ -268,18 +276,49 @@ local unsigned long crc32_generic(crc, buf, len)
...
+	uint32_t regs[4]; // output: eax, ebx, ecx, edx
+	__cpuid( regs, leaf );
+	if (leaf == 1) {
+		ecx = regs[2];
+	#else
+ 	if (__get_cpuid(leaf, &eax, &ebx, &ecx, &edx)) {
+	#endif
+		if ((ecx & crc_bit_PCLMUL) != 0)
+			cpu_has_pclmul = 1;		
+		//printf("leaf=%d, eax=0x%x, ebx=0x%x, ecx=0x%x, edx=0x%x, cpu_has_pclmul %d\n", leaf, eax, ebx, ecx, edx, cpu_has_pclmul);
```
Comment (vkrasnov) [inline crc32.c]: please remove

File: crc32.c [hunk 4]
```diff
@@ -288,7 +327,10 @@ uLong crc32(crc, buf, len)
 {
     if (len < PCLMUL_MIN_LEN + PCLMUL_ALIGN  - 1)
       return crc32_generic(crc, buf, len);
-
+	#ifndef SKIP_CPUID_CHECK
+	if (!has_pclmul())
+      return crc32_generic(crc, buf, len);
+    #endif
```
Comment (vkrasnov) [inline crc32.c:331]: Not aligned with the `#ifndef` above it

File: deflate.c [hunk 5]
```diff
@@ -545,7 +545,7 @@ int ZEXPORT deflateTune(strm, good_length, max_lazy, nice_length, max_chain)
  * upper bound of about 14% expansion does not seem onerous for output buffer
  * allocation.
  */
-uint64_t ZEXPORT deflateBound(strm, sourceLen)
+uLong ZEXPORT deflateBound(strm, sourceLen)
```
Comment (vkrasnov) [inline deflate.c]: why is this changed?

**Comment by neurolabusc** [comment]
@vkrasnov thanks for your careful review. I believe my latest commit includes all your suggestions.

**Review by kornelski — APPROVED** [review]

**Comment by vkrasnov** [comment]
@neurolabusc thank you very much for the contribution!
